[Infrastructure] LocalNet on Kubernetes

okdas opened the pull request:
## Description

Implements LocalNet that is powered by Kubernetes. 

## Issue

~~pocket-operator issue: https://github.com/pokt-network/pocket-operator/issues/7~~ not using operator currently, just utilizing Kubernetes directly.

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Major breaking change
- [x] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

Introducing new dependencies to run localnet developer environment
- [Tilt](https://tilt.dev/)
- Kubernetes - the workloads controlled by Tilt and pocket-operator run on local kuberenetes. Potential choices are:
  - Docker Desktop
  - Rancher Desktop
  - other options: https://docs.tilt.dev/choosing_clusters.html

## Testing

- [ ] `tilt up` should bring the whole stack online.
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [x] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

okdas committed:
tilt + PO + localnet

okdas committed:
--wip-- [skip ci]

okdas committed:
--wip-- [skip ci]

okdas committed:
observability stack

okdas committed:
--wip-- [skip ci]

okdas committed:
grafana is auto provisioned with observability stack

ManInWeb3 reviewed (approved):
lgtm
Great job

Olshansk commented:
> lgtm Great job

Before we merge this in, @okdas can you show (gif, video, demo, instructions) of it working? 

Also, please make sure to follow the LocalNet test steps to make sure it works.

okdas commented:
@Olshansk of course, I'm adding the instructions! :)

okdas committed:
--wip-- [skip ci]

okdas committed:
stabilizing tilt

Olshansk commented:
> @Olshansk of course, I'm adding the instructions! :)

@okdas Lmk when this is ready for a spin.

okdas committed:
--wip-- [skip ci]

okdas committed:
more changes to grafana, tilt and docs

okdas commented:
> > @Olshansk of course, I'm adding the instructions! :)
> 
> @okdas Lmk when this is ready for a spin.

Still in progress, but if you want to check it out locally and provide feedback/see if you catch any issues please make sure to follow the doc and also utilize the `dk-localnet-pocket-operator` branch from pocket-operator (the second repo is needed - more info in docs).

okdas committed:
more docs

okdas committed:
let tilt ignore the same stuff as git, so it wont get triggered by vendor/mock/etc changes

okdas committed:
note about changing pocket-operator code

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+
+### How to change configuration files
+
+LocalNet resources are managed by pocket-operator, so if you'd like to change a config file (e.g. genesis file), you can do it by modifying the corresponding CRD in `pocket-operator` repo.
```
I think I should expand on that, but this should probably go into the pocket-operator documentation and linked here.

okdas commented:
@Olshansk please lmk if you can try this on your machine.

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+
+* [tilt](https://docs.tilt.dev/install.html)
+* Kubernetes cluster ([different options available](https://docs.tilt.dev/choosing_clusters.html)), please make sure to run 1.23 version or older, until [this issue is resolved](https://github.com/zalando/postgres-operator/issues/2098) - otherwise Postgres DB is not going to be provisioned.
+  * `kubectl` is also required to be installed and configured to access the cluster, but that should happen automatically when you install kubernetes cluster locally.
```
Can we remove this line given the point above?

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+* [tilt](https://docs.tilt.dev/install.html)
+* Kubernetes cluster ([different options available](https://docs.tilt.dev/choosing_clusters.html)), please make sure to run 1.23 version or older, until [this issue is resolved](https://github.com/zalando/postgres-operator/issues/2098) - otherwise Postgres DB is not going to be provisioned.
+  * `kubectl` is also required to be installed and configured to access the cluster, but that should happen automatically when you install kubernetes cluster locally.
+  * `helm` is required to template the yaml manifests for the observability stack. Here are the instructions on how to install it: https://helm.sh/docs/intro/install/.
```
Let me know what you think of this style of writing for development docs. Personally, I don't think people like to read - just `copy-pasta` as fast as possible

```suggestion
  * `helm` - required to template the yaml manifests for the observability stack.
  	* Installation instructions: : https://helm.sh/docs/intro/install/.
```

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+* Kubernetes cluster ([different options available](https://docs.tilt.dev/choosing_clusters.html)), please make sure to run 1.23 version or older, until [this issue is resolved](https://github.com/zalando/postgres-operator/issues/2098) - otherwise Postgres DB is not going to be provisioned.
+  * `kubectl` is also required to be installed and configured to access the cluster, but that should happen automatically when you install kubernetes cluster locally.
+  * `helm` is required to template the yaml manifests for the observability stack. Here are the instructions on how to install it: https://helm.sh/docs/intro/install/.
+* pocket kubernetes operator codebase in `../pocket-operator` directory, relative to the pocket v1 codebase. If you don't have it, tilt will try to pull it from the `main` branch on first launch.
```
Why not make `pocket-operator a git submodule`?

Then we won't have a weird setup and it'll be easier to automate.

okdas replied:
From my experience, submodules are a huge PITA when used on repos where multiple people are collaborating. Though I agree, that would give us an ability to tie configuration files versions and logic from pocket-operator to the v1 repo. I just fear that managing a submodule might become a nightmare.

I mentioned that above - I won't be surprised if we remove pocket-operator from LocalNet and will use Kubernetes directly (with Tilt controlling the whole lifecycle). So I guess it doesn't matter if we use submodules or not - they are not going to be a deciding factor if operator sticks for LocalNet or not. :) 

Do you see any issues with using submodule? I can swap Tiltfile code with submodule.

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+
+### Running the localnet
+
+Start the LocalNet with `make localnet_up` command - this will start tilt. Tilt will prompt to press `space` to open a browser window with the UI where you can see the logs of all services running (including the operator, observability stack, etc.). You can also open the UI by going to `http://localhost:10350/` in your browser.
```
Alternative format for this section. Wdyt?
--- 

```
make localnet_up
```

The developer can then view the logs of all the services running (including the operator, observability stack, etc.) by either:
1. Pressing `space`
2. Going to [localhost:10350](http://localhost:10350/)

okdas replied:
Love it! ❤️

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+
+### Stopping / cleaning up the resources
+
+`make localnet_down` will stop the localnet and clean up all the resources, except the postgres operator (in case you have other databases provisioned with it, the database itself will be destroyed).
```
If you do the following (for this command and others), the copy command will show up on github!

```
make localnet_down
```

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+As the workloads run in Kubernetes, you can see and modify any resources on your local kubernetes by a tool of your choice (k9s, Lens, VSCode extension, etc.) - just be mindful that tilt and pocket-operator will change the resources back eventually, so you might want to disable tilt or turn off pocket-operator to make changes that you want to test.
+
+We provide following make targets:
+`make localnet_shell` - opens a shell in the pod that has `client` cli available. It gets updated automatically whenever the code changes.
```
These are great comments and I think we should improve the Makefile to contain them. That's where developers will be looking!

Olshansk commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+`make localnet_shell` - opens a shell in the pod that has `client` cli available. It gets updated automatically whenever the code changes.
+`make localnet_client_debug` - opens a `client debug` cli. It allows to interact with blockchain, e.g. change pace maker mode, reset to genesis, etc. It gets updated automatically whenever the code changes.
+
+### How does it work?
```
Wdyt of adding a mermaid diagram? https://mermaid-js.github.io/mermaid/#/

Stopping here for tonight. Will continue reviewing tomorrow

okdas replied:
That is a great idea, working on it.

okdas replied:
I initially built a larger diagram that was hard to make sense of, so I only kept the part with interaction between tilt, operator, and containers. Because that is the most tricky part.

Olshansk reviewed (changes requested):
I got through about half of the review and almost got it running locally! I don't think this will be ready to merge by EOD tomorrow but hopefully, it's a start.

Do you think you'd be able to help me look into why it's not working on my machine? I'm attaching a couple of photos by haven't investigated it at all.



<img width="1608" alt="Screenshot 2022-12-12 at 11 12 04 PM" src="https://user-images.githubusercontent.com/1892194/207250378-b8292c8b-660f-44b2-a7de-6ebce39b7afb.png">

<img width="1608" alt="Screenshot 2022-12-12 at 11 12 07 PM" src="https://user-images.githubusercontent.com/1892194/207250362-99d2d5e3-bab1-4281-8716-cb039b9df2fd.png">

okdas commented:
> Do you think you'd be able to help me look into why it's not working on my machine? I'm attaching a couple of photos by haven't investigated it at all.

I have an idea why this might happen. Since all four validators and pocket-v1-cli-client tilt resources are unavailable with an error that binary is missing, I suspect golang is unable to compile them.
`pocket: Watch & Compile` and `debug client: Watch & Compile` tilt resources are responsible for building them, and they would show errors. Possibly related to vendor dir.

Usually I have `Alerts on top` checkbox selected to see the issues. Optionally, you probably would have seen this error if clicked on `All Resources`.

Quick question: do you think this is important to have resource name convention consistent? E.g. on your screenshot you can see some names are kebab case, but some (like examples above that compile a binary) contain spaces, etc.

okdas commented on .gitignore:
```diff
@@ -60,4 +60,9 @@ rpc/server.gen.go
\ No newline at end of file
+build/config/gen*.json
+
+# Binary downloads and manipulations by helm
```
This is not related to operator, but helm instead. Helm is being used here to download third-party Open Source helm charts for observability part, and it basically downloads zip files. I don't want us to commit binaries, so that is why this is included in gitignore.

okdas commented on .tiltignore:
```diff
@@ -0,0 +1,64 @@
+# Binaries for programs and plugins
```
I don't believe there is ([documentation piece related to this](https://docs.tilt.dev/file_changes.html#where-ignores-come-from)). Plus, you make a good point of removing vendor `dir` from it below, so here they are different.

okdas commented on .tiltignore:
```diff
@@ -0,0 +1,64 @@
+*.out
+
+# Dependency directories (remove the comment below to include it)
+vendor/
```
Good catch – yes, I believe we should rebuild the binary in this case. I want to keep the mocks here though, because they are being changed during test runs. I've been experimenting with using tilt to run tests continuously – but changing files by tests during tests results in an infinite loop of constantly running tests… :)

okdas commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+# LocalNet on Kubernetes
```
hackerman!

okdas commented on Makefile:
```diff
@@ -474,6 +474,28 @@ gen_genesis_and_config:
 clear_genesis_and_config:
 	rm build/config/gen.*.json
 
+.PHONY: localnet_up
```
I want to exercise an idea of doing these checks in Tiltfile. Obviously, need to do a `tilt` check in make target though. Will address this.

okdas committed:
Update Makefile

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas commented on Makefile:
```diff
@@ -474,6 +474,28 @@ gen_genesis_and_config:
+	kubectl exec -it deploy/pocket-v1-cli-client -- client debug
+
+.PHONY: localnet_shell
+## Starts debug client (allows to control LocalNet)
```
`localnet_client_debug` prompts you directly into `debug client`, `localnet_shell` shells you into a container that has `pocket` and `client` cli available. Do you think this doesn't make sense or just need to be documented better (in .md and make target)?

okdas committed:
Update Makefile

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas commented on Makefile:
```diff
@@ -474,6 +474,28 @@ gen_genesis_and_config:
+
+.PHONY: localnet_shell
+## Starts debug client (allows to control LocalNet)
+localnet_shell:
```
With `localnet_shell` you get into a pod with `pocket` and `client` CLI available. Both binaries are automatically updated when code is changed in the repo. How do I make this more obvious – document it in .md files and make target?

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,114 @@
+load('ext://helm_resource', 'helm_resource', 'helm_repo')
+load('ext://namespace', 'namespace_create')
+
+# TODO: add check k8s is 1.23, 1.24+ is supported once https://github.com/zalando/postgres-operator/issues/2098 is resolved.
```
I've added an automated check now that fails with error asking to downgrade.

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,114 @@
+load('ext://helm_resource', 'helm_resource', 'helm_repo')
```
In this PR I introduce a helm chart `build/localnet/observability-stack` that pulls open source helm charts for observability (prometheus, grafana, loki, etc.) and connects them all together.

okdas replied:
The chart now also handles postgres, hence it has been renamed to just `dependencies`.

okdas committed:
Update Makefile

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas commented on Makefile:
```diff
@@ -474,6 +474,28 @@ gen_genesis_and_config:
+	tilt down
+	# kubectl delete --all --namespace=default pocketvalidators
+	kubectl get sts -n default --no-headers=true | awk '/v1-validator/{print $1}' | xargs kubectl delete -n default  sts
+
```
I suppose so, alternatively I suspect a change might be done in a Tiltfile to do this. I've looked into this only briefly and decided to change a make target.

This is only necessary because of our set-up with operator, which controls only some resources, when tilt controls the rest of them.

okdas commented on app/client/cli/debug.go:
```diff
@@ -50,6 +47,9 @@ var (
 		PromptShowLatestBlockInStore,
 	}
+
+	defaultConfigPath  = getEnv("CONFIG_PATH", "build/config/config1.json")
```
Hm, this certainly works because we set `CONFIG_PATH` for `client` pod. https://github.com/pokt-network/pocket/blob/dk-localnet-pocket-operator/build/localnet/cli-client.yaml#L24

okdas commented on build/localnet/observability-stack/dashboards/README.md:
```diff
@@ -0,0 +1,3 @@
+# Grafana dashboards for localnet
+
+Include json files in this directory to have them automatically imported into Grafana.
```
@Olshansk there is one now, imported from existing local net with slight adjustments because log lines have changed.

okdas commented on build/localnet/observability-stack/Chart.yaml:
```diff
@@ -0,0 +1,5 @@
+name: observability-stack
+version: 0.0.1
+appVersion: v0.0.1
+description: "Helm chart for deploying the observability stack. This chart is mostly used for managing observability dependencies and wiring them up together."
```
`Chart.yaml` is always in the root of the helm chart, so `build/localnet/observability-stack`.

okdas commented on build/localnet/restart.sh:
```diff
@@ -0,0 +1,25 @@
+# Further reading:
+# https://docs.tilt.dev/live_update_reference.html#restarting-your-process
+#
+# Usage:
```
Short answer is yes if we continue to use operator. I've tried different approaches but the standard, recommended by `tilt`, way does not work without additional manual manipulations. And this is documented in the link included in .sh file: https://docs.tilt.dev/live_update_reference.html#restarting-your-process


There are a few exceptions to the above; the restart_process extension doesn’t work for:
* ...
*    CRDs


To be frank, I am still torn if we should use operator here. tilt works great with Kubernetes, and even operators too (minus some quirks like this one or having to cleanup resources managed by operator), but until we add more logic into operator (such as dynamically scaling actors) it is not as beneficial comparing to just going with `tilt` directly. And `tilt` directly would also allow to scale LocalNet dynamically, this is just a different approach to solve the same problem.

okdas commented on build/localnet/start.sh:
```diff
@@ -0,0 +1,44 @@
+# Further reading:
+# https://docs.tilt.dev/live_update_reference.html#restarting-your-process
+#
+# Usage:
```
Same answer! :)

okdas commented on Makefile:
```diff
@@ -474,6 +474,28 @@ gen_genesis_and_config:
 	rm build/config/gen.*.json
 
+.PHONY: localnet_up
+## Starts localnet with all dependencies (basically, `tilt up`)
```
Good idea, but it doesn't seem `tilt` is meant to be installed via `go install`.
e.g. `go install "github.com/tilt-dev/tilt/cmd/tilt@v0.30.13"` fails, and it is not a supported way to install tilt. This would be convinient though..

okdas replied:
It is not supported indeed. https://github.com/tilt-dev/tilt/issues/5334#issuecomment-1121173647

Olshansk replied:
Why not just put `curl -fsSL https://raw.githubusercontent.com/tilt-dev/tilt/master/scripts/install.sh | bash` in the `install_cli_deps` target? It's call `cli_deps`, not `go_cli_deps` :)

okdas replied:
@Olshansk I guess you're right we can do that, though, I've been avoiding `curl *.sh | bash` due to security concerns. Okay. Let do this.

okdas committed:
requested changes so far

Olshansk commented:
@okdas Please re-request a review once it's ready for round2

okdas committed:
more requested changes

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
some requested changes

okdas commented:
@Olshansk  Working on mermaid diagram(s), and possibly a gif demo, the rest is ready for round2.

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
add mermaid diagram

okdas commented on docs/development/localnet-with-kubernetes.md:
```diff
@@ -0,0 +1,69 @@
+### Dependencies
+
+* [tilt](https://docs.tilt.dev/install.html)
+* Kubernetes cluster ([different options available](https://docs.tilt.dev/choosing_clusters.html)), please make sure to run 1.23 version or older, until [this issue is resolved](https://github.com/zalando/postgres-operator/issues/2098) - otherwise Postgres DB is not going to be provisioned.
```
Tilt now checks this

okdas commented on build/localnet/observability-stack/values.yaml:
```diff
@@ -0,0 +1,74 @@
+grafana:
```
This only is related to grafana and related software (such as prometheus or loki), in that case what should the video capture? grafana UI with working dashboard?

okdas committed:
clarify location of the config files

okdas committed:
--wip-- [skip ci]

okdas commented:
I still experience instabilities with LocalNet, so I'm going to reduce a number of moving parts:
- remove Postgres Operator and provision PostgresDB with static StatefulSet and credentials;
- if still going to experience issues, will (temporarily) remove `pocket-operator` from LocalNet, submitting validator manifests directly to K8s via Tilt. Once we have a stable LocalNet, which I don't want to delay further, I can iterate to introduce back pocket-operator.

okdas committed:
localnet without operator

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
latest config

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
fully working operatorless localnet

okdas committed:
auto update now works too

okdas commented:
Okay, I've got it working reliably without any operators. Will get ready to demo both approaches and explain the problem on example.

okdas committed:
--wip-- [skip ci]

okdas committed:
bump documentation

okdas committed:
change config a bit

okdas commented on runtime/genesis.go:
```diff
@@ -11,6 +12,7 @@ import (
 func parseGenesis(genesisJSONPath string) (g *genesis.GenesisState, err error) {
 	data, err := os.ReadFile(genesisJSONPath)
 	if err != nil {
+		err = fmt.Errorf("%w while reading %s", err, genesisJSONPath)
```
this and below:
Without additional information I was not able to figure out which file contained a mistake.

okdas commented:
@Gustavobelfort do you have a chance to review and test this on your machine?

okdas committed:
this is not needed anymore

Gustavobelfort commented:
> @Gustavobelfort do you have a chance to review and test this on your machine?

Will do so today!

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
can scale validators automatically

okdas commented:
@Olshansk I added automatic scaling in this PR since it was not yet reviewed and depends on tilt. Here is how scaling is done: https://github.com/pokt-network/pocket/blob/a778f2c8d531cc4b85711a38b5651478172dc7ac/docs/development/localnet-with-kubernetes.md#scaling-actors-on-localnet

Now, validators are actually not participating because they are not mentioned in the genesis file. From what I understand, we can pre-populate genesis with max amount of validators we think we will ever run on LocalNet realistically (10? 50?), or wait until peer-to-peer is done. Is there another option?

deblasis commented:
@okdas 

> Now, validators beyond the original four are actually not participating because they are not mentioned in the genesis file. From what I understand, we can pre-populate genesis with max amount of validators we think we will ever run on LocalNet realistically (10? 50?), or wait until peer-to-peer is done. Is there another option?

This is coming with #416 (which is essentially a basic bootstrapping + advertising of churn changes that will keep the address books updated) as an interim solution but, you got it, P2P is gonna be key here.

okdas commented on build/localnet/private-keys.yaml:
```diff
@@ -0,0 +1,1014 @@
+  # val3: "5db3e9d97d04d6d70359de924bb02039c602080d6bf01a692bad31ad5ef93524c16043323c83ffd901a8bf7d73543814b8655aa4695f7bfb49d01926fc161cdb" # 67eb3f0a50ae459fecf666be0e93176e92441317
+  # val4: "6fd0bc54cc2dd205eaf226eebdb0451629b321f11d279013ce6fdd5a33059256b2eda2232ffb2750bf761141f70f75a03a025f65b2b2b417c7f8b3c9ca91e8e4" # 6f66574e1f50f0ef72dff748c3f11b9e0e89d32a
+
+  # New keys that start with a number that corresponds to a number of a validator. First two letters after the digits correspond to a type of actor. aa, being the first actor transitioning to this, is the validator.
```
I realized that `aa` is probably not the best choice considering we already have `ActorType` attribute. So I'll regenerate the addresses to reflect actor type we already settled on.

okdas committed:
change keys

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
also replace pubkeys

okdas commented:
re: ci errors - I likely forgot to change priv/pub key somewhere - for some reason vscode didn't show all the places for replacements, I'll double-check with ack/grep tomorrow first thing. The rest is good for review.

okdas committed:
complete changing the keys // tests passed

okdas committed:
bring bck state hashes

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
wip

okdas committed:
wip

okdas committed:
add/change changelogs

Olshansk committed:
Updated gitinore for Tilt

Olshansk committed:
Made the Makefile help command work

Olshansk committed:
Renamed the readme file and fixed default formatting

Olshansk committed:
Update error messages

Olshansk committed:
Update error messages

okdas committed:
create localnet_config.yaml when it doesn exist

Olshansk committed:
WIP commit - review + removing $ everywhere

Olshansk committed:
Update all the documentation

Olshansk commented on build/localnet/README.md:
````diff
@@ -0,0 +1,165 @@
+## Code Structure
+
+```bash
+build/localnet # TODO_IN_THIS_COMMIT
````
@okdas please update the `TODO_IN_THIS_COMMIT` comments

Olshansk commented on build/localnet/README.md:
```diff
@@ -0,0 +1,165 @@
+
+### Docker Desktop
+
+# TODO_IN_THIS_COMMIT
```
Can you please update this based on our offline conversation?

okdas replied:
What exactly do you want to see here? Docker Desktop, as any other software, might not work (frankly, I was never happy with it on mac). If you don't want to troubleshoot issues, you can resort to other software.

In this readme file, I included a list of distributions I've tested this LocalNet on, linked a page on tilt.dev website describing all possible options (basically - any cluster/distribution will work, as long as Kubernetes is available).

Olshansk commented on build/localnet/README.md:
```diff
@@ -0,0 +1,165 @@
+When starting a k8s LocalNet, `localnet_config.yaml` is generated (with default configs) in the root of the repo if doesn't already exist.
+
+The config file can be modified to scale the number of actors up/down. As long as `localnet_up` is running, the changes should be automatically applied within seconds.
+
```
Could you add a GIF or video here?

okdas replied:
uploading it here, will include in readme file.

https://user-images.githubusercontent.com/4950477/216490690-8ac4c16a-25e1-4202-b2e5-03215030c82c.mp4

okdas replied:
I don't think there's a way to embed videos in readme, huh..

Olshansk commented on build/localnet/README.md:
```diff
@@ -0,0 +1,165 @@
+# Kubernetes LocalNet <!-- omit in toc -->
+
+This guide shows how to deploy a LocalNet using [pocket-operator](https://github.com/pokt-network/pocket-operator).
```
@okdas I updated most of this file while testing so please review.

okdas replied:
Looks good, I also added some changes (mostly requested ones). Please lmk if you have questions.

deblasis committed:
feat(localnet): initcontainer to wait for postgres

deblasis reviewed (commented):
I took the liberty of adding this https://github.com/pokt-network/pocket/pull/354/commits/f75fde249df89e04e430b8c37f6eb90773fbcb05 to avoid the errors when the validators spin up before postgres. Feel free to revert it as you see fit.

Other than that:
- I have added a comment regarding my setup when installing dependencies.
- I am thinking about a way to stake and unstake validators when they come online/go offline during the scale-up scale-down operations. My idea is to write a simple program that interacts with `client-go` to react to k8s events and stake/unstake accordingly (obviously I'd have to change the genesis so that the validators have POKT to stake and also provide access to the keys of the validators to this program, parsing the `private-keys.yaml` I guess).  
- Currently we can scale-up or scale-down by modifying the yaml file but if I implement the program above, we could have an api call do the operation instead, not sure if this should reside inside the debug client but I am interested in your views about it. 

@Olshansk, @okdas: Do you see any issues and/or do you have any suggestions regarding the above? 

If it sounds good at a high level, I'll proceed with creating an issue that highlights the design more thoroughly.

If I manage to do that + the KISS peer discovery on top of this **great work** we'd have something pretty cool IMHO

deblasis commented on Makefile:
```diff
@@ -101,6 +101,7 @@ install_cli_deps: ## Installs `protoc-gen-go`, `mockgen`, 'protoc-go-inject-tag'
 	go install "github.com/golang/mock/mockgen@v1.6.0" && mockgen --version
 	go install "github.com/favadi/protoc-go-inject-tag@latest"
 	go install "github.com/deepmap/oapi-codegen/cmd/oapi-codegen@v1.11.0"
+	curl -fsSL https://raw.githubusercontent.com/tilt-dev/tilt/master/scripts/install.sh | bash
```
Also, I guess that `helm` CLI binary is a requirement that should be considered in the `install_cli_deps` as well

okdas replied:
> this command fails with

Oh no, I was sure tilt install script can be installed on Linux because the same script is offered for both macOS and Linux. https://docs.tilt.dev/install.html

Looking at the script, even on Linux, they check if `brew` CLI is available. That is probably because [Homebrew supports Linux](https://docs.brew.sh/Homebrew-on-Linux). When `brew` cli is available, they try to install `tilt` with it. Do you use `brew` a lot on your Linux WSL2 set up? It is possible your local `brew` set up might be out of date, have you tried to run the commands mentioned to fix this?

It would be easier to install `tilt` with `go install` like we do with other binaries, but unfortunately this is [not supported](https://github.com/tilt-dev/tilt/issues/5334).


> Also, I guess that `helm` CLI binary is a requirement that should be considered in the `install_cli_deps` as well

Thanks for mentioning this, I added a helm install script to that make target. 🫡

okdas committed:
also install helm

okdas commented:
> I took the liberty of adding this [f75fde2](https://github.com/pokt-network/pocket/commit/f75fde249df89e04e430b8c37f6eb90773fbcb05) to avoid the errors when the validators spin up before postgres. Feel free to revert it as you see fit.

Looks good, thank you! For local development environment this is perfect. I did something like this inside Tiltfile initially (waiting for postgres to be available and only then deploy validators), but then decided to let Kubernetes restart the process until postgres is up. I do like your approach because it would not let the validator Pods to be marked as unhealthy when postgres is not available yet. Thank you!

>     * I am thinking about a way to stake and unstake validators when they come online/go offline during the scale-up scale-down operations. My idea is to write a simple program that interacts with `client-go` to react to k8s events and stake/unstake accordingly (obviously I'd have to change the genesis so that the validators have POKT to stake and also provide access to the keys of the validators to this program, parsing the `private-keys.yaml` I guess).

This is where operator would shine.. Because operator process runs inside the cluster, has all necessary permissions, already subscribed on all changes AND provides an ability for reconciliation (e.g. if stake command fails - it can be executed again). 

>     * Currently we can scale-up or scale-down by modifying the yaml file but if I implement the program above, we could have an api call do the operation instead, not sure if this should reside inside the debug client but I am interested in your views about it.

The reason why I went with the config file is because this is something I used to when managing resources via GitOps pattern. I understand that we do not commit that config file, but the idea is similar - you mutate the config, and the setup automagically provisions everything to the state described in the yaml config file. So it would be easy to write a make target or a script that changes the config file, which would trigger a scale-up/scale-down event.

What is a use case for an API call? I ask because this is what Operator would provide potentially, and I would love to know how you would use an API call for scaling validators and other actors.

Operator is built on top of [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) which is an amazing SDK to control resources in cluster. We already have that, and I currently work on a high-level roadmap for operator, so we can include this kind of changes into that roadmap! :)

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,100 @@
+    print('Updating localnet_config.yaml')
+    local('cat - > localnet_config.yaml', stdin=encode_yaml(localnet_config))
+
+# List of directories Tilt watches to trigger a hot-reload on changes
```
Instead of listing the directories in readme I am going to provide a link to the Tiltfile to that line.

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,100 @@
+])
+
+# # Exposes postgres port to 5432 on the host machine.
+# k8s_resource(new_name='postgres',
```
I would keep it in case developer wants to connect to the postgres cluster with their own tooling (dbeaver, etc.) - that requires exposing a port. I changed the comment to reflect this is an example. Do you think this makes sense?

okdas commented on build/localnet/cli-client.yaml:
```diff
@@ -0,0 +1,71 @@
+apiVersion: apps/v1
```
Done, I couldn't find any code that uses this though. I only find a mention of `GetIsClientOnly()` in `shared/CHANGELOG.md`. Is this something going to be used in future?

okdas commented on build/localnet/cli-client.yaml:
```diff
@@ -0,0 +1,71 @@
+      volumes:
+        - name: config-volume
+          configMap:
+            # should client have it's own config?
```
Haha, I put this comment there before I added a default config for all validators. All of them use just one config file with values changed via environment variables only when necessary.

okdas commented on build/localnet/configs.yaml:
```diff
@@ -0,0 +1,325 @@
+apiVersion: v1
```
Are you referring to `config.json`? If you look at this file you left a comment on, you find that JSON is still there and was not converted into YAML. 

`ConfigMap` allows to mount values described in it to a file. For example, in our `v1-validator-default-config` `ConfigMap` we have `config.json` as a key and a string (what happens to be a json output!) as a value. 

In a Pod volumes section, we reference [v1-validator-default-config](https://github.com/pokt-network/pocket/blob/b46d2d3597f50d8de71c875a83f7f3fe50060a53/build/localnet/v1-validator-template.yaml#L84) `ConfigMap` that allows us to mount that `config.json` as a file. 

A great benefit in this system is we only need to maintain one default config file for all validators, and overriding the parameters via environment variables only for parameters where it is necessary.

okdas commented on build/localnet/private-keys.yaml:
```diff
@@ -0,0 +1,1023 @@
+# Private keys on LocalNet have addresses that start with specific pattern:
+# NNNXX, where NNN is a number of a validator and XX is a type of actor.
```
done

okdas committed:
requested changes so far

okdas commented on build/localnet/private-keys.yaml:
```diff
@@ -0,0 +1,1023 @@
+  # val3: "5db3e9d97d04d6d70359de924bb02039c602080d6bf01a692bad31ad5ef93524c16043323c83ffd901a8bf7d73543814b8655aa4695f7bfb49d01926fc161cdb" # 67eb3f0a50ae459fecf666be0e93176e92441317
+  # val4: "6fd0bc54cc2dd205eaf226eebdb0451629b321f11d279013ce6fdd5a33059256b2eda2232ffb2750bf761141f70f75a03a025f65b2b2b417c7f8b3c9ca91e8e4" # 6f66574e1f50f0ef72dff748c3f11b9e0e89d32a
+
+  # New keysthat start with a number that corresponds to a number of a validator. First two letters after the digits correspond to a type of actor. aa, being the first actor transitioning to this, is the validator.
```
Sorry, I should've removed that! Initially I didn't account for `ActorType` initially and used `aa` for validators.  Once I realized we already have an existing `ActorType` protobuf, I switched to that (validator == `04`) but forgot to remove these few lines.

okdas commented on build/localnet/private-keys.yaml:
```diff
@@ -0,0 +1,1023 @@
+# Private keys on LocalNet have addresses that start with specific pattern:
```
I used https://github.com/okdas/pocket-vanitygen, pushed the modified version of it to a separate branch and attached a link in comments for future reference - we are going to need to generate more keys for other actors/networks.

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,132 @@
+load("ext://restart_process", "docker_build_with_restart")
+
+# Create localnet config file from defaults, and if some default configuration doesn't exist in it - populate with default values
+localnet_config_defaults = {"validators": {"count": 4}}
```
Added a link to a `Tiltfile` where defaults are at.

okdas committed:
more requested changes

okdas commented on build/localnet/README.md:
```diff
@@ -0,0 +1,165 @@
+├── network.yaml # TODO_IN_THIS_COMMIT
+├── private-keys.yaml # TODO_IN_THIS_COMMIT
+├── v1-validator-template.sh # TODO_IN_THIS_COMMIT
+└── v1-validator-template.yaml # TODO_IN_THIS_COMMIT
```
They belong to different dirs, so I suggested a different structure, please check this file again.

okdas committed:
update links

okdas commented on Tiltfile:
```diff
@@ -0,0 +1,132 @@
+# Helm charts are used for the observability stack (Grafana, Prometheus, Postgres)
```
It was not straight forward due to `root_dir` changed but seem to work. Please check new dir structure.

okdas committed:
include demo

Olshansk commented:
> https://github.com/pokt-network/pocket/commit/f75fde249df89e04e430b8c37f6eb90773fbcb05

I really like the `until nc -z` trick. Brilliant.

> * I am thinking about a way to stake and unstake validators when they come online/go offline during the scale-up scale-down operations. My idea is to write a simple program that interacts with `client-go` to react to k8s events and stake/unstake accordingly (obviously I'd have to change the genesis so that the validators have POKT to stake and also provide access to the keys of the validators to this program, parsing the `private-keys.yaml` I guess).

@okdas My understanding is this:

1. We couldn't get pocket-operator working on LocalNet (for now)
2. We'll use pocket-operator in DevNet
3. The current configuration scales the number of validators (e.g. services) but doesn't actually trigger the stake/unstake command based on k8s events

The flow @deblasis described might literally let us scale (and discover) validators in a local environment next week, so I'd support going in that approach unless I'm missing something?

Olshansk commented on build/localnet/README.md:
```diff
@@ -132,34 +157,29 @@ If force triggering an update didn't work, try the following:
+│   ├── cli-client.yaml # Pod that has the latest binary of the pocket client. Makefile targets run CLI in this pod.
+│   ├── configs.yaml # Location of the config files (default configs for all validators and a genesis file) that are shared between all actors
+│   ├── network.yaml # Networking configuration that is shared between different actors, currently a Service that points to all validators
+│   └── private-keys.yaml # Pre-generated private keys for everything
```
```suggestion
│   └── private-keys.yaml # Pre-generated private keys with a semantic format for easier development
```

Olshansk commented on build/localnet/Tiltfile:
```diff
@@ -3,25 +3,29 @@ load("ext://helm_resource", "helm_resource", "helm_repo")
 load("ext://namespace", "namespace_create")
 load("ext://restart_process", "docker_build_with_restart")
 
+tiltfile_dir = os.path.dirname(config.main_dir)
```
Thank you for doing this! I think it's going to make the codebase much easier to understand

Olshansk commented on build/localnet/Tiltfile:
```diff
@@ -3,25 +3,29 @@ load("ext://helm_resource", "helm_resource", "helm_repo")
+    local("cat - > " + localnet_config_path, stdin=encode_yaml(localnet_config))
 
-# List of directories Tilt watches to trigger a hot-reload on changes
+# List of directories Tilt watches to trigger a hot-reload on changes. Potentially can be replaced with a list of excluded directories.
```
```suggestion
# List of directories Tilt watches to trigger a hot-reload on changes.
# CONSIDERATION: This can potentially can be replaced with a list of excluded directories.
```

Olshansk commented on build/localnet/Tiltfile:
```diff
@@ -105,19 +115,19 @@ COPY bin/client-linux /usr/local/bin/client
 )
 
-# # Exposes postgres port to 5432 on the host machine.
+# An example of how to expose postgres port to 5432 on the host machine in case developer wants to connect to postgres directly
```
```suggestion
# Uncomment in order to expose postgres port to 5432 on the host machine in case the developer wants to connect to postgres directly
```

Olshansk replied:
Question: Would this conflict with the postgres that's managed by Tilt?

okdas replied:
This is to expose the postgres database managed by tilt. But it might conflict with any other postgres that runs on your machine. I mentioned how to avoid that in my recent edit.

Olshansk commented on build/localnet/templates/v1-validator-template.yaml.tpl:
```diff
@@ -69,6 +71,15 @@ spec:
               subPath: genesis.json
             - name: validator-storage
               mountPath: /validator-storage
+      initContainers:
```
I like this :)

Olshansk reviewed (changes requested):
A small set of minor NITs and comments and we're good to go!

The only major point I wanted to discuss is why we shouldn't move forward with the approach outlined by Alessandro.

okdas commented on build/localnet/README.md:
```diff
@@ -67,6 +80,11 @@ When starting a k8s LocalNet, `localnet_config.yaml` is generated (with default
 
+[Watch demo](https://user-images.githubusercontent.com/4950477/216490690-8ac4c16a-25e1-4202-b2e5-03215030c82c.mp4)
+
+![Watch demo](https://user-images.githubusercontent.com/4950477/216490690-8ac4c16a-25e1-4202-b2e5-03215030c82c.mp4)
+
```
It actually looks nice on the github markdown preview! You have a video and a link. But yeah, we can remove it.

okdas commented on build/localnet/README.md:
```diff
@@ -93,17 +111,24 @@ Open a `client debug` cli. It allows to interact with blockchain, e.g. change pa
+### Addresses and keys on LocalNet
 
-[tilt](https://tilt.dev/) reads the [`Tiltfile`](../../Tiltfile), where LocalNet configs are specified. `Tiltfile` is written in [Starlark](https://github.com/bazelbuild/starlark), a dialect of Python.
+You can find private keys and addresses for all actors in the [private-keys.yaml](./manifests/private-keys.yaml) file. They have been pre-generated and follow a specific pattern - they start with pre-determined numbers for easier troubleshooting and debugging.
```
I did, it is described in private-keys.yaml file.

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/Tiltfile

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Update build/localnet/Tiltfile

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
clarify postgres ports

okdas commented:
> The only major point I wanted to discuss is why we shouldn't move forward with the approach outlined by Alessandro.

We totally should! I love what @deblasis did!

I think I've addressed all of your suggestions, let me resolve conflicts and it'll be ready for another round.

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
fix changelogs

okdas committed:
oops, this got lost!

Olshansk commented on build/localnet/README.md:
```diff
@@ -112,8 +112,11 @@ make localnet_client_debug
+Addresses begin with `YYYXX` number, where `YYY` is a number of an actor and `XX` is [a type of actor](../../shared/core/types/proto/actor.proto#L7). For example:
+*  `420043b854e78f2d5f03895bba9ef16972913320` is a validator #420.
+*  `66603bc4082281b7de23001ffd237da62c66a839` is a fisherperson #666.
+*  `0010297b55fc9278e4be4f1bcfe52bf9bd0443f8` is a service node (a servicer) #001.
```
Just FYI: We're going to rename all `Service Node` like instances to `Servicer` soon

Olshansk commented on build/localnet/README.md:
```diff
@@ -115,8 +112,11 @@ make localnet_client_debug
 
-Addresses begin with `YYYXX` number, where `YYY` is a number of an actor and `XX` is [a type of actor](../../shared/core/types/proto/actor.proto#L7). For example, `420043b854e78f2d5f03895bba9ef16972913320` is a validator #420.
-
+Addresses begin with `YYYXX` number, where `YYY` is a number of an actor and `XX` is [a type of actor](../../shared/core/types/proto/actor.proto#L7). For example:
+*  `420043b854e78f2d5f03895bba9ef16972913320` is a validator #420.
```
```suggestion
Addresses begin with `YYYXX` number, where `YYY` is a number of an actor and `XX` is [a type of actor](../../shared/core/types/proto/actor.proto#L7).

The current mapping for `XX` is:
* `01` - Application
* `02` - Servicer
* `03` - Fisherman
* `04` - Validator

For example:
*  `420043b854e78f2d5f03895bba9ef16972913320` is a validator #420.
```

Olshansk reviewed (approved):
Just one minor suggestion, but otherwise, LGTM!

okdas committed:
Update build/localnet/README.md

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

okdas committed:
Merge branch 'main' into dk-localnet-pocket-operator

okdas committed:
change date on changelog

okdas committed:
redundant changelog
